### PR TITLE
Fix six silent wrong answers in formula, bootstrap and stepwise handling (#226, #273, #275, #277, #278, #279)

### LINKED ISSUE #226: Vector interface does not translate Surv() censoring codes, so left- and interval-censored rows are silently refitted as right-censored
Found by the `r-reviewer` pass over the #213 fix, and verified at `2cf7a35` (pre-existing, not introduced by that change).

## What happens

`AGENTS.md` states the rule: this package codes censoring `-1` left, `0` right, `1` event, `2` interval, while `survival::Surv()` under `type = "interval"` uses `0`/`1`/`2`/`3` for right/event/left/interval. It also says the formula path translates in `.hzr_parse_formula()` and **the vector path does not**.

`R/hazard_api.R:556-557` unclasses a `Surv` status column without translating it, so those codes reach the likelihood unchanged.

## Reproducer

```r
ph <- list(early = hzr_phase("cdf"), late = hzr_phase("constant"))
s  <- survival::Surv(c(1,2,3,4,5,6), c(1,0,1,1,0,1), type = "left")

f <- hazard(time = c(1,2,3,4,5,6), status = s, dist = "multiphase",
            phases = ph, fit = FALSE, objective = "sas")
f$data$status
#> 1 0 1 1 0 1        <- rows 2 and 5 are LEFT-censored; stored as right-censored
```

The same `Surv()` through the formula interface stops correctly:

```r
d <- data.frame(tt = c(1,2,3,4,5,6), ev = c(1,0,1,1,0,1))
hazard(survival::Surv(tt, ev, type = "left") ~ 1, data = d, dist = "multiphase",
       phases = ph, fit = FALSE, objective = "sas")
#> Error: objective = "sas" does not support left-censored rows (status == -1) ...
```

No error, no warning, a fit over the wrong censoring model. This is the defect `AGENTS.md` records the package having already shipped once (`Surv(type = "left")` read left-censored rows as right-censored).

`type = "interval"` has the same shape with an extra wrinkle: status `3` (interval) is matched by no branch of `.hzr_logl_multiphase()`, so those rows contribute nothing to the likelihood at all.

## Consequence for #213

The entry guard added in #213 (`.hzr_check_sas_data()`) checks the codes it is given, so on the vector path a `Surv()` left-censored row is invisible to it. Its `@details` and the `NEWS.md` entry both say so explicitly rather than implying coverage it does not have. Closing this issue would let that caveat be removed.

## Suggested fix

Translate at the point of the unclass in `R/hazard_api.R:556-557`, using the same mapping `.hzr_parse_formula()` applies, so both interfaces agree. Worth extracting that mapping into one internal helper both call, since the asymmetry has now produced defects more than once.

A `Surv` object carries its `type` in `attr(s, "type")`, so the translation can be driven off that rather than guessed from the observed code values — guessing would misread a right/event-only vector.

### LINKED ISSUE #273: hazard(Surv(t, d) ~ .) includes the response columns as covariates
**Silent wrong fit.** `hazard(Surv(t, d) ~ ., data = df)` puts the response's own columns into the covariate design:

```r
library(TemporalHazard); data(avc)
d <- na.omit(avc)[, c("int_dead", "dead", "age", "mal")]
f <- hazard(survival::Surv(int_dead, dead) ~ ., data = d, dist = "weibull",
            theta = c(mu = 0.01, nu = 0.5, 0, 0, 0, 0))
colnames(f$data$x)
#> "int_dead" "dead" "age" "mal"      -- the follow-up time and the event indicator are covariates
```

With a `theta` sized for `age` and `mal` only, the fit stops with "non-conformable arguments", which does not name the cause. With a `theta` sized for all four, it fits the **outcome as a predictor**, with no error or warning.

**Cause.** `.hzr_parse_formula()` (`R/formula-helpers.R`) rebuilds the RHS as `formula(paste("~", deparse(rhs)))` and calls `model.matrix()` on the whole `data`. On a one-sided formula, `.` expands to every column, including the ones the `Surv()` LHS uses. In a two-sided formula, `model.frame()` would exclude them.

Other consequences:
- `predict(newdata = )` then demands `int_dead` and `dead` columns.
- On `fix/predict-newdata-phase-columns` it trips the new `time`-covariate check (#270) when a response column is named `time`.

**Fix direction.** Expand `.` against the data minus the LHS variables, e.g. `stats::terms(formula, data = data)` on the full two-sided formula, then drop the response; or refuse `.` with a clear error. Either way, add a test that `colnames(fit$data$x)` excludes the `Surv()` variables.

Found by `r-reviewer` on `fix/predict-newdata-phase-columns` (review of #270). Present on `main` (`8a26c0e`); `.hzr_parse_formula()` has not changed there in this respect.

### LINKED ISSUE #275: hazard(): a phase-scoped global term such as ~ constant(age) is silently dropped
**Silent wrong model.** A phase-scoped term in the global formula, of the form `phase_name(var)`, is dropped without a warning:

```r
library(TemporalHazard); data(avc); d <- na.omit(avc)
f <- hazard(survival::Surv(int_dead, dead) ~ constant(age), data = d, dist = "multiphase",
  phases = list(early = hzr_phase("cdf", t_half = 0.5, nu = 1, m = 1, fixed = "shapes"),
                constant = hzr_phase("constant")), fit = TRUE)
names(f$fit$theta)
#> "early.log_mu" "early.log_t_half" "early.nu" "early.m" "constant.log_mu"   -- no age coefficient
f$fit$covariate_counts
#> 0 0
```

The fit converges as a covariate-free model and reports nothing.

**Cause.** `hazard()` (`R/hazard_api.R`, ~lines 534-548) detects a phase-scoped RHS with `.hzr_formula_has_phase_scope()` (`R/stepwise-formula.R`) and replaces it with `~ 1`. The comment there says covariate routing is "resolved inside .hzr_optim_multiphase()", but no code reads the phase-scoped terms back and routes them to their phases. So they are simply lost.

**Fix direction.** Either:
- route each `phase_name(...)` term into that phase's `formula` (combining it with any `hzr_phase(formula =)` already given), with a test that `names(fit$fit$theta)` includes `constant.age`; or
- if the syntax is not meant to be supported, stop with an error that points to `hzr_phase(formula =)`.

Also check the stepwise path, which builds phase-scoped formulas (`R/stepwise-formula.R`).

Found by the fifth `r-reviewer` pass on `fix/predict-newdata-phase-columns` and reproduced at `b3c3b7f`. It predates #270; `main` (`8a26c0e`) has the same code.

### LINKED ISSUE #277: hzr_phase(formula = ~ .) puts the Surv() response columns into the phase design
`hzr_phase(formula = ~ .)` puts the `Surv()` response columns into the phase's design matrix, so the outcome is fitted as a phase covariate. This is the phase-formula counterpart of #273, which fixed only the global formula.

Reproduced on `main` at `1e1bd27`, from a clean `git archive` export:

```r
library(survival)
a <- na.omit(avc[, c("int_dead", "dead", "age", "mal")])
fit <- hazard(Surv(int_dead, dead) ~ 1, data = a, dist = "multiphase",
  phases = list(early = hzr_phase("cdf", t_half = 0.15, nu = 1.4, m = 1,
                                  fixed = "m", formula = ~ .),
                constant = hzr_phase("constant")),
  fit = TRUE, control = list(n_starts = 1L, conserve = FALSE))
colnames(fit$fit$x_list$early)
#> "int_dead" "dead" "age" "mal"
fit$fit$converged   #> TRUE
fit$fit$objective   #> -37.13479
```

The vector interface does the same: `hazard(time = int_dead, status = dead, data = a, ...)` with the same phases also gives an early design of `int_dead dead age mal`.

The fit converges with no error. The phase formula is expanded by `model.frame(ph$formula, data)` in `R/likelihood-multiphase.R`, `R/score-test.R` and multiphase `predict()`, each of which sees every column of `data`.

**This blocks v1.2.11.**

### LINKED ISSUE #278: hzr_bootstrap() does not resample a formula variable that is not in data, and reports success
`hzr_bootstrap()` resamples the rows of the fit's `data`, but a formula variable that is not a column of `data` is not resampled. Each replicate pairs the original covariate vector with resampled rows, so the interval is wrong. Every replicate is reported as a success, with no warning.

Reproduced on `main` at `1e1bd27`, from a clean `git archive` export:

```r
library(survival)
a <- na.omit(avc[, c("int_dead", "dead", "age", "mal")])
zz_glob <- a$age   # the same values as the `age` column, but outside `data`
fz <- hazard(Surv(int_dead, dead) ~ zz_glob, data = a, dist = "weibull",
             theta = c(0.3, 1, 0), fit = TRUE)
fc <- hazard(Surv(int_dead, dead) ~ age, data = a, dist = "weibull",
             theta = c(0.3, 1, 0), fit = TRUE)
bz <- hzr_bootstrap(fz, n_boot = 40, seed = 1)
bc <- hzr_bootstrap(fc, n_boot = 40, seed = 1)
```

| Fit | Estimate | `n_success` / `n_failed` | Bootstrap 95% CI |
|---|---|---|---|
| `~ zz_glob` (workspace variable) | -0.00587 | 40 / 0 | (-0.00147, 0.00129) |
| `~ age` (column) | -0.00587 | 40 / 0 | (-0.01194, -0.00239) |

The workspace-variable interval excludes its own estimate. Nothing in the result says anything went wrong.

On `main` a function-local variable fails at the fit with "object not found". On the `fix/vector-surv-codes` branch (#273), the formula's own environment is used, so a function-local variable fits too and gives the same wrong bootstrap.

The fix agreed with the maintainer is to refuse: `hzr_bootstrap()` stops, and names every right-hand-side variable, global or phase formula, that is not a column of `data`.

**This blocks v1.2.11.**

### LINKED ISSUE #279: hzr_stepwise() on a ~ . base fit sees no terms and reports zero steps
`hzr_stepwise()` on a base fit written as `Surv(...) ~ .` sees no terms, so it reports a finished screen of zero steps. That cannot be told apart from an honest "nothing met `slstay`".

The cause is `.hzr_formula_rhs_terms()` in `R/stepwise-formula.R`. It calls `stats::terms(formula)` without `data`, which errors on `.`, and a `tryCatch()` turns the error into `character()`.

Reproduced on `main` at `1e1bd27`, from a clean `git archive` export. On `main`, `~ .` also carries the response columns (#273), so `theta` is sized for that design:

```r
library(survival)
a2 <- na.omit(avc[, c("int_dead", "dead", "age", "mal")])
set.seed(1); a2$noise <- rnorm(nrow(a2))
fd <- hazard(Surv(int_dead, dead) ~ ., data = a2, dist = "weibull",
             theta = c(0.3, 1, rep(0, 5)), fit = TRUE)
TemporalHazard:::.hzr_formula_rhs_terms(Surv(int_dead, dead) ~ .)
#> character(0)
sw <- hzr_stepwise(fd, data = a2, direction = "backward",
                   criterion = "wald", slstay = 0.2, trace = FALSE)
NROW(sw$steps)
#> 0
```

The same model with its terms written out drops `noise` (p = 0.385 in the #273 review run).

The fix agreed with the maintainer is to refuse: stop, and ask for the base model's terms to be written out.

**This blocks v1.2.11.**

### PR DESCRIPTION
**Pull request by ehrlinger**
Six defects found in the v1.2.11 run-up, all converged or finished runs that gave a wrong answer with no error. Each is its own commit with its tests, merged with `main` at `e497a4a`.

Closes #226
Closes #273
Closes #275
Closes #277
Closes #278
Closes #279

## What changes

- **#226** A `survival::Surv()` passed as `status` on the vector interface is now read by its `type`, as the formula interface always did. It used to take the second column unchanged, so a left-censored row was fitted as right-censored, and `"interval"` / `"counting"` read `time2` / `stop` as status codes. Both interfaces now go through `.hzr_surv_response()`.
- **#273** `.` in a `hazard()` formula now means every column the `Surv()` term does not use, as in `survival::coxph()`. It used to put the response in the design, so the outcome was fitted as a predictor and the fit converged.
- **#275** A multiphase formula that names a phase as a function (`~ constant(age)`) is now refused. The term used to be stripped with the rest of the right-hand side and never reached its phase.
- **#277** `hzr_phase(formula = ~ .)` is now written out once in `hazard()`, before fitting, against `data` without the `Surv()` variables. The likelihood, the score test, `predict()` and the stored spec all see explicit terms.
- **#278** `hzr_bootstrap()` now refuses a fit whose response, covariates or select-mode `scope` use a per-row variable that is not a column of `data`. Such a variable was never resampled: on `avc`, 40 of 40 replicates succeeded, with no warning, and the interval excluded its own estimate. Constants outside `data` (`pi`, a cutoff, a knots vector) are still allowed.
- **#279** `hzr_stepwise()` now refuses a `~ .` base fit before printing anything. It used to read no terms and report a finished screen of zero steps.

## Breaking changes (calls that used to run now stop)

- `hazard(..., dist = "multiphase")` with a phase called as a function in the global formula (#275). Move the covariates into `hzr_phase(formula = ~ var)`.
- `hzr_bootstrap()` on a fit that uses a per-row variable outside `data` (#278). Add it to `data` and refit.
- `hzr_stepwise()` with a `~ .` base formula, a `scope = ~ .`, or a multiphase base whose global formula uses `.` while a phase inherits it (#279). Write the terms out.
- `.` in a phase formula on the vector interface (`time =`, `status =`), where no `Surv()` term says which columns hold the response (#277).
- `.` beside other terms when `data` holds only the response columns (#273).
- A `time`, `time_lower` or `time_upper` that disagrees with a `Surv` passed as `status` (#226).

**Estimates change** for any earlier fit that used `.` in the global or a phase formula (#273, #277), and so does the length of `theta` a `~ .` single-distribution fit needs.

## Gates (final tree, `384b668`)

- `devtools::document()`: `man/` and `NAMESPACE` clean on a second run (`git diff --exit-code`).
- `lintr::lint_package()`: 0 lints. `spelling::spell_check_package()`: no errors.
- `NOT_CRAN=true devtools::test()`, with `/Volumes/qhsstudies` mounted from start to end: **4674 passed, 0 failed, 6 skipped, 0 errors**. Both `test-repeated-events-parity.R` (5 tests) and `test-repeated-events-renewal-parity.R` (1 test) ran and passed. The skips:
  - `test-bhdead-sas-parity.R` x3: bh.dead SAS fixture not available;
  - `test-parity-c-binary.R` x1: the legacy binary exits with status 126;
  - `test-weights-sas-parity.R` x2: `weighted-avc-fractional.rds` not found.
- `R CMD check --as-cran` with the manual, from a clean `git archive` export: **Status OK (0/0/0)**, 228 s overall (tests 103 s, vignettes 42 s), tarball 3.20 MB, no `CLAUDE`/`AGENTS` files.
- Every fix was written test-first, and a mutation check reverted each fix and confirmed its tests fail.
- `r-reviewer` ran over each round (#273, #275, #277-#279, their follow-ups, and the merge with the doc commit). Each finding was reproduced before acting on it. The ones fixed:
  - the `.` edge cases;
  - the non-multiphase `dist` guard;
  - bootstrap false refusals of constants;
  - the response and select-mode `scope` checks, including the multiphase scope environment (`384b668`);
  - the multiphase `~ .` stepwise exemption;
  - NEWS figures that did not reproduce.

  The merge resolutions and the doc commit came back clean.

## Known, non-blocking

- `inst/dev/PLAN-hessian-analytic-multiphase.md` still shows the old `~ early(...)` global-formula syntax that #275 now refuses. It is a historical plan, excluded from the tarball.
- A phase named `log` makes `log(age)` in the global formula trigger the #275 refusal; `base::log(age)` fits. The helper's roxygen now says so.
- `hzr_bootstrap()`'s check works on variable names, so indirection such as `I(lst$a)` or `get("zz")` escapes it, and a constant vector whose length happens to equal the row count is refused. Both are rare; a value-based check would be a redesign.
- Pre-existing, not from this branch: adding a formula to a phase in a multiphase stepwise run drops the global covariate that phase inherited; and `document()` warns about an unresolved link at `R/likelihood-loglogistic.R:239`, which is identical on `main`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ehrlinger** [comment]
Back to draft: Copilot's review (13:08Z on `384b668`) has items that look valid and are not yet fixed. Mostly they concern the #278 bootstrap guard resolving response and `scope` symbols in the wrong environment, plus a vector-interface `x =` matrix that is never resampled, and a `@param formula` sentence the new `Surv` contract contradicts. The fixes are coming on this branch. It will be marked ready again once they are pushed and every item is answered.

File: R/diagnostics.R [hunk 1]
```diff
@@ -1488,6 +1486,86 @@ hzr_bootstrap <- function(object, n_boot = 200L, fraction = 1.0,
...
+    } else if (is.character(scope) && length(scope) > 0L) {
+      all.vars(stats::reformulate(scope))
+    }
+    phase_formulas <- Filter(function(f) inherits(f, "formula"),
+                             lapply(object$spec$phases, function(ph) {
+                               ph$formula
+                             }))
+    outside <- c(
+      if (length(stored) == 3L) {
+        outside_in(all.vars(stored[[2L]]), environment(.hzr_parse_formula))
```
[the remark this answers is not in the document]
Comment (ehrlinger) [reply R/diagnostics.R:1525]: Declined, with evidence. The namespace lookup already finds the global binding, and switching to `base_env` would reintroduce a silent case.

- `environment(.hzr_parse_formula)` is the package namespace. Its parent chain runs namespace -> imports -> base namespace -> `R_GlobalEnv` -> search path, so `get0("tt_ws", envir = asNamespace("TemporalHazard"), inherits = TRUE)` returns the global binding (checked: `TRUE`).
- On this branch, `hazard(Surv(tt_ws, dead) ~ age, data = a, ...)` with `tt_ws` in the global environment, then `hzr_bootstrap(fit)`, stops with "uses 'tt_ws', which is not a column". The test `a response variable outside \`data\` is refused` in `tests/testthat/test-bootstrap-data-columns.R` covers exactly your example and passes.
- The namespace is used on purpose. `.hzr_parse_formula()` evaluates the `Surv()` term in `data`, then its own frame and the namespace, never in the formula's environment. A caller-local variable therefore never reaches the fit ("object not found"), and a caller-local variable with the same name as a global one is ignored by the fit. Looking the response up in `base_env` would read that local instead of the vector the fit used. The test `the response is checked where the fit evaluates it` fails if the lookup is switched to `base_env` (mutation-checked).
- Line 1663 is a comment in the vector-interface block and does no lookup.

No code change for this item. The other bootstrap scope-environment findings from this review are fixed in 61f2597; see the PR comment.

File: R/hazard_api.R [hunk 2]
```diff
@@ -640,9 +653,35 @@ hazard <- function(formula = NULL,
...
-  # Convert Surv object status to numeric if needed (after formula parsing)
+  # A Surv object passed as `status` is read exactly as the formula path reads
+  # it (#226). Its codes are not this package's, and under "interval" and
+  # "counting" its second column is not the status at all, so taking that
+  # column unchanged fitted left-censored rows as right-censored. The Surv
+  # defines the bounds it carries; `time` and any bound the caller also gave
+  # must agree with it rather than be silently overridden.
   if (inherits(status, "Surv")) {
-    status <- unclass(status)[, 2L]
+    resp <- .hzr_surv_response(status)
```
[the remark this answers is not in the document]
Comment (ehrlinger) [reply R/hazard_api.R:664]: Declined, with evidence. `length()` of a `Surv` is the number of observations, not the number of matrix elements, so `length(status) == length(time)` compares n with n.

survival registers an S3 method:

```r
> getS3method("length", "Surv")
function (x)
nrow(x)
<environment: namespace:survival>
```

On 4-row objects, `length()` is 4 for every type:

| `Surv` type | `length()` |
|---|---|
| right (`Surv(t, e)`) | 4 |
| left (`type = "left"`) | 4 |
| interval (`type = "interval2"`) | 4 |
| counting (`Surv(t, t2, e)`) | 4 |

survival is always loaded when a `Surv` exists, because building one requires it.

A `Surv` passed as `status` therefore reaches the #226 block. `tests/testthat/test-vector-surv-status.R` passes `Surv` objects through `hazard(time =, status =)` for all four types, and passes. See "Surv '%s' gives the same %s fit on both interfaces", "an interval2 Surv, left-censored row included, matches the formula", and "a Surv status whose times disagree with 'time' or a bound stops".

No code change for this item.

**Comment by ehrlinger** [comment]
These are Copilot's four suppressed items from the 13:08Z review of `384b668`. I reproduced each one by running code before changing anything. The fixes are in **61f2597**. Each has a test that fails on `384b668`, and each passed a mutation check.

**1. `R/diagnostics.R:1535`: scope names were checked only in `base_env`.** Valid, fixed.
- A single-distribution `scope = ~ zl + mal`, built in a frame the refit cannot see, bootstrapped with `n_success = 3`.
- `zl` could not be refitted. After one up-front warning it dropped silently out of every replicate's candidates: the summary listed only `mal`, and `n_uncomputable_replicates` was 0.
- A scope formula's variables are now also checked in `environment(scope)`, so this call is refused.

**2. `R/diagnostics.R:1549`: a scope for a phase with no formula was checked only in the namespace.** Valid, fixed. `scope = list(constant = ~ zl)`, with `zl` local to the frame that built the scope, ran with `n_success = 2`. Each phase scope is now also checked in `environment(sc)`, as well as where its refit resolves it (the phase formula's environment, or the namespace for a phase with no formula).

**3. `R/diagnostics.R:1665`: a vector-interface `x =` was not resampled.** Valid, fixed by refusing it, in both fixed and select mode. It predates #283.
- **Your exact example** (no `data`) was already refused, on this branch and on `main`, but with a garbled message that names the vectors `'NA', 'NA'` because `n_obs` comes from the missing `data`.
- **The silent case is `data =` plus a direct `x`.** `hazard(data = a, time = int_dead, status = dead, x = as.matrix(a["age"]))` gave 20/20 successes and an `age` interval of (-0.0016, 0.0013) around an estimate of -0.0059. The formula path gives (-0.0146, -0.0029). `main` at `e497a4a` gives the identical result.
- **Select mode had the same defect, only worse.** The base refits reused the original `x`, and the candidate refits dropped it, so the `age` terms vanished from every replicate while all of them reported success.
- **The fix.** A vector-interface fit with a direct `x` is now refused before `set.seed()`, so a refused call leaves the caller's random number stream alone. A stale comment that claimed such an `x` was rebuilt each replicate is corrected.
- **Not changed.** The underlying `.hzr_refit_with_scope()` drop of `x` predates #283 and is noted separately. The garbled `'NA'` message for a vector fit with no `data` and no `x` is left as it is: it fails loud, and it also predates #283.

**4. `R/hazard_api.R:139`: `@param formula` told callers to pass numeric codes with `status`.** Valid, fixed. It now says that a plain `status` vector takes this package's codes, and that a `Surv` passed as `status` is translated the same way as on the formula path.

The two inline threads are declined; the evidence is on each thread.

**Gates on `61f2597`**
- `document()` is clean on a second run; lint 0; spelling clean.
- `NOT_CRAN=true` suite, with `/Volumes/qhsstudies` mounted from start to end: 4679 passed, 0 failed, 6 skipped. The skips are the bh.dead fixture x3, the legacy C binary x1 and the weighted-parity fixture x2. `test-repeated-events-parity.R` and `test-repeated-events-renewal-parity.R` both ran and passed.
- `R CMD check --as-cran` with the manual, from a clean `git archive` export: Status OK (0/0/0), 230 s overall (tests 103 s, vignettes 42 s), tarball 3.20 MB, no `CLAUDE`/`AGENTS` files.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
